gpjax/abstractions.py

```diff
@@ -192,6 +194,80 @@ def step(carry, iter_num__and__key):
     return inf_state
 
 
+def fit_natgrads(
+    stochastic_vi: StochasticVI,
+    params: Dict,
+    trainables: Dict,
+    transformations: Dict,
+    train_data: Dataset,
+    moment_optim,
+    hyper_optim,
+    key: PRNGKeyType,
+    batch_size: int,
+    n_iters: Optional[int] = 100,
+    log_rate: Optional[int] = 10,
+) -> Dict:
+    """This is a training loop for natural gradients. See Salimbeni et al. (2018) Natural Gradients in Practice: Non-Conjugate Variational Inference in Gaussian Process Models
+
+    We begin with an initalise natural gradient step to tighten the ELBO for hyperparameter optimisation. There after, each iteration comprises a hyperparameter gradient step followed by natural gradient step to avoid a stale posterior.
+
+    Args:
+        stochastic_vi (StochasticVI): The stochastic variational inference algorithm to be used for training.
+        params (Dict): The parameters for which we would like to minimise our objective function with.
+        trainables (Dict): Boolean dictionary of same structure as 'params' that determines which parameters should be trained.
+        transformations (Dict): The transformations to be applied to the parameters.
+        train_data (Dataset): The training dataset.
+        batch_size(int): The batch_size.
+        key (PRNGKeyType): The PRNG key for the mini-batch sampling.
+        n_iters (int, optional): The number of optimisation steps to run. Defaults to 100.
+        log_rate (int, optional): How frequently the objective function's value should be printed. Defaults to 10.
+    Returns:
+        InferenceState: A dataclass comprising optimised parameters and training history.
+    """
+
+    hyper_state = hyper_optim.init(params)
+    moment_state = moment_optim.init(params)
+
+    nat_grads_fn, hyper_grads_fn = natural_gradients(
+        stochastic_vi, train_data, transformations
+    )
```

Comment:
These lines seem to be the only ones that require the `stochastic_vi` object. Can we simply move this outside of the training loop and then just make the function accept two objective_fn; one for the hyperparams and one for the natgrads?